### DISCUSSION TITLE
It's complicated to mock INavigator on unit tests

### DISCUSSION THREAD

**Discussion by takla21** [unoplatform/uno.extensions#795] [Ideas]
We were able to add some unit testing on uno.extensions project, however it's complicated to mock it on our viewmodels as it's driven by extension method while also strongly depending on other services that somehow the navigator references.

Here's a list of thing we had to do in order to configure our test with navigation (for instance with Moq).
```cs
var mockInstance = new Mock<IInstance<IServiceProvider>>();

// Mock RouteResolver to inject expected destination viewmodel or add default behavior so we can complete the navigation execution.
 var routeSolver = new Mock<IRouteResolver>();
var route = default(RouteInfo);
 if (destinationViewModel != null)
 {
      route = new RouteInfo(Path: destinationViewModel.Name);
}

 routeSolver.Setup(x => x.FindByViewModel(destinationViewModel ?? It.IsAny<Type>(), It.IsAny<INavigator>()))
       .Returns(route);

var mockServiceProvider = new Mock<IServiceProvider>();
mockServiceProvider.Setup(x => x.GetService(typeof(IRouteResolver)))
      .Returns(routeSolver.Object);

 mockInstance.SetupGet(x => x.Instance)
     .Returns(mockServiceProvider.Object);

// This makes sure that the mock instance implement both INavigator & IInstance<IServiceProvider>
var mockNavigator = mockInstance.As<INavigator>();
```

This whole thing to test logic with this method `INavigator.NavigateViewModelAsync`.

This should not be that complicated to have unit testing with navigations.

**Comment by takla21** [comment]
Also forgot to mention that the workaround that I've added is definitely prone to break when we would probably have viewmodel with more complicated navigation (e.g. dialogs, routes, etc...)

**Reply by nickrandolph** [reply]
We'll have to explore options on how to make this experience better. One thought would be for extensions to provide a way to get a mock navigator but this would assume a particular mocking framework.

**Reply by takla21** [reply]
I actually think there could be two way to approach this:
- The hard one : Simplify the actual implementation of `INavigator` so it does not include `IInstance<IServiceProvider>`
- The easiest one: Extensions should provide another navigator implementation that simply can be used for integration tests. For instance, on this framework https://github.com/nventive/Chinook.Navigation, there's a concept of a blindnavigation that essential is a simplification of the navigation service implementation that can be used on .net standard projects.
See https://github.com/nventive/Chinook.Navigation#ready-for-integration-testing
